pkg_resources/__init__.py

```diff
@@ -2200,8 +2199,12 @@ def resolve_egg_link(path):
 
 register_finder(importlib.machinery.FileFinder, find_on_path)
 
-_declare_state('dict', _namespace_handlers={})
-_declare_state('dict', _namespace_packages={})
+_namespace_handlers: Dict[
+    type, Callable[[object, str, str, types.ModuleType], Optional[str]]
+] = {}
+_declare_state('dict', _namespace_handlers=_namespace_handlers)
+_namespace_packages: Dict[Optional[str], List[str]] = {}
+_declare_state('dict', _namespace_packages=_namespace_packages)
```

Comment:
To be honest, given this approach of pre-declaring the global variables, it seems that `_declare_state` could be simplified to something like:

```python
_declare_state(vartype: str, varname: str) -> None:
    _state_vars[varname] = vartype
```

In turn that would allow us to simplify:

```python
_distribution_finders: Dict[
    type, Callable[[object, str, bool], Iterable["Distribution"]]
] = {}
_declare_state('dict', '_distribution_finders')
```

OR if we want to have a "oneliner"/"decorator-style-ish":

```python
_declare_state(vartype: str, varname: str, value: T) -> T:
    _state_vars[varname] = vartype
    return value

...

_distribution_finders: Dict[
    type, Callable[[object, str, bool], Iterable["Distribution"]]
] = _declare_state('dict', '_distribution_finders', {})

# it is a pity Python don't have an equivalent of decorators for assignment statements...
# (so that we wouldn't need to write the name of the variable as an argument).
```

But that is a refactoring that might require further discussion and maybe a separated PR, so just posting here for the sake of brain-storming.

Reply:
TODO: I hope I remember this suggestion in the future :P

Reply:
I don't remember at which python version this becomes supported (with `from __future__ import annotations`), but one could even do:
```py
_distribution_finders = _declare_state(
    'dict',
    '_distribution_finders',
    dict[type, Callable[[object, str, bool], Iterable["Distribution"]]](),
)
```

Reply:
https://github.com/pypa/setuptools/pull/4346